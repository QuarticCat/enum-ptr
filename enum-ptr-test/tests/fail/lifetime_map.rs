#![allow(dead_code)]

use enum_ptr::{Compact, EnumPtr};

#[derive(EnumPtr)]
#[repr(C, usize)]
enum Foo<'a, 'b> {
    A(&'a i32),
    B(&'b i32),
}

fn test_map_ref() {
    let foo: Compact<_> = Foo::A(&1).into();
    foo.map_ref(|f| match f {
        Foo::A(r) | Foo::B(r) => r,
    });
}

unsafe fn test_map_mut() {
    let mut foo: Compact<_> = Foo::A(&1).into();
    foo.map_mut(|f| match f {
        Foo::A(r) | Foo::B(r) => r,
    });
}

fn main() {}
