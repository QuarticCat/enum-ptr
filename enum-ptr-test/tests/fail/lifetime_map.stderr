error: lifetime may not live long enough
  --> tests/fail/lifetime_map.rs:15:34
   |
14 |     foo.map_ref(|f| match f {
   |                  -- return type of closure is &'2 &i32
   |                  |
   |                  has type `&'1 Foo<'_, '_>`
15 |         Foo::A(r) | Foo::B(r) => r,
   |                                  ^ returning this value requires that `'1` must outlive `'2`

error: lifetime may not live long enough
  --> tests/fail/lifetime_map.rs:22:34
   |
21 |     foo.map_mut(|f| match f {
   |                  -- return type of closure is &'2 mut &i32
   |                  |
   |                  has type `&'1 mut Foo<'_, '_>`
22 |         Foo::A(r) | Foo::B(r) => r,
   |                                  ^ returning this value requires that `'1` must outlive `'2`
